jenkins_node_scanner.py

```diff
@@ -0,0 +1,157 @@
+import sys
+import time
+
+import jenkins
+from prometheus_client import start_http_server, Counter, Gauge, Summary
+
+
+JENKINS_API_LATENCY = Summary(
+    'jenkins_api_latency_seconds',
+    'Response latency of Jenkins API (seconds)',
+    ['api_method'],
+)
+
+JENKINS_API_EXCEPTIONS = Counter(
+    'jenkins_api_exceptions',
+    'Exceptions encountered accessing the Jenkins API.',
+    ['api_method'],
+)
+
+GLOBAL_EXCEPTIONS = Counter(
+    'node_scraper_exceptions',
+    'Unhandled top-level exceptions.',
+)
+
+NODES_FOUND = Gauge(
+    'nodes_found',
+    'Nodes found in a scan.',
+    ['jenkins_master'],
+)
+
+UNPARSEABLE_NODES_FOUND = Gauge(
+    'unparseable_nodes_found',
+    'Nodes found in a scan whose ips could not be parsed.',
+    ['jenkins_master'],
+)
+
+
+@contextmanager
+def ignore_exceptions_except_exit():
+    """Ignores all exceptions except KeyboardInterrupt and SystemExit."""
+    try:
+        yield
+    except (KeyboardInterrupt, SystemExit):
+        raise
+    except:
+        logging.exception('Unexpected error, continuing anyway.')
+
+
+def get_args():
+    """Configure arguments and parse them."""
+    parser = argparse.ArgumentParser(
+        description='Write Jenkins nodes to a file.',
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument('url',
+                        help='Full URL to the Jenkins master, such as http://jenkins.')
+    parser.add_argument('output_file',
+                        help='Filename to write the node list.')
+    parser.add_argument('--username',
+                        help='Jenkins username.')
+    parser.add_argument('--password',
+                        help='Jenkins password.')
+    parser.add_argument('--exclude-regex', default='^master$',
+                        help='Exclude any nodes matching this regex.')
+    parser.add_argument('--timeout', default=60, type=int,
+                        help='Timeout in seconds for the Jenkins call.')
+    parser.add_argument('--period', default=60, type=int,
+                        help='How many seconds to wait between scans.')
+    parser.add_argument('--prometheus-port', default=8000, type=int,
+                        help='Port on which to expose prometheus metrics.')
+    parser.add_argument('--target-port', default=9100, type=int,
+                        help='Target port to be scraped (e.g. 9100 for node_exporter).')
+    parser.add_argument('-v', '--verbose', action='store_true',
+                        help='Enable more detailed logging.')
+    return parser.parse_args()
+
+
+def get_node_ip(master, node):
+    """Returns the ip of a node."""
+    with JENKINS_API_EXCEPTIONS.labels('get_node_config').count_exceptions():
+        with JENKINS_API_LATENCY.labels('get_node_config').time():
+            node_config = master.get_node_config(node['name'])
+    if 'hudson.plugins.swarm.SwarmSlave' not in node_config:
+        logging.error('Unrecognized node type: %s', node['name'])
```

Comment:
I would replace this with `logging.info` or `logging.warning`, since we can easily encounter this error with the master executor (in some cases).